### ISSUE TITLE
Bugs in Haskell-level Re-exports

### ISSUE THREAD

**Issue by crockeea** [haskell/haddock#584]
I've just been trying to re-export modules with Haddock, but I ran into several annoying issues:

1. If you export renamed modules as in 
   ```
   module Foo (module X) where
   import module A as X
   import module B as X
   ```
   documentation for `A` and `B` do not appear in the documentation for `Foo`.
2. If you export _unrenamed_ modules, documentation for `A` and `B`_does_ appear in docs for `Foo`:
   ```
   module Foo (module A, module B) where
   import module A
   import module B
   ```
3. If you import a module hiding some items, and then re-export the module, the hidden items still appear in the documentation:
   ```
   module Foo (module A, module B) where
   import module A hiding (a)
   import module B
   ```
   In this example, documentation for `a` would appear in the docs for `Foo`, even though `a` is not exported from `Foo`.

(All of this assumes that modules `A` and `B` are hidden in the cabal file, so that their contents should be dumped into `Foo` in an ideal world.)

It doesn't seem that (1) is addressed in [the docs](https://www.haskell.org/haddock/doc/html/ch03s04.html#idm140354810861024). In particular, the modules are re-exported "wholly and without hiding qualifiers", and the docs mention nothing about renamed modules. 

Point (2) is precisely what I expect/want to happen, so nothing to change there. 

Point (3) may be addressed in the footnote [2] in the docs (can't quite tell), but this feature is pretty useless with the current behavior. The only option is for me to explicitly list what items I want to export, which is error-prone and difficult to maintain.

**Comment by capitanbatata** [comment]
We're also affected by this. I thought it was a problem in the way we were using Haddock, so I asked [a question at SO](https://stackoverflow.com/questions/46452424/haddock-generating-documentation-for-multiple-re-exported-modules-through-the-s).

**Comment by gbaz** [comment]
should be resolved by #688
